### PR TITLE
rpc: Improve scantxoutset response and help message

### PR DESCRIPTION
**Pull request by promag**
The new response keys `height` and `bestblock` allow the client to know at what point the scan took place.

The help message now has all the response keys (`result` and `txouts` were missing) and it's improved a bit. Note that `searched_items` key is renamed to `txouts`, considering `scantxoutset` is marked experimental.

### CODE REVIEW

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -2293,15 +2297,19 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
         int64_t count = 0;
         std::unique_ptr<CCoinsViewCursor> pcursor;
+        CBlockIndex* tip;
         {
             LOCK(cs_main);
             ::ChainstateActive().ForceFlushStateToDisk();
             pcursor = std::unique_ptr<CCoinsViewCursor>(pcoinsdbview->Cursor());
-            assert(pcursor);
+            tip = ::ChainActive().Tip();
+            assert(pcursor && tip);
```
Comment (MarcoFalke) [inline src/rpc/blockchain.cpp]: Could leave them in separate lines, so that an assert failure is more verbose and helpful?

Comment (promag) [reply src/rpc/blockchain.cpp]: Done.

**Comment by promag** [comment]
@jonasschnelli mind taking a look?

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -2293,15 +2297,20 @@ UniValue scantxoutset(const JSONRPCRequest& request)
         g_scan_progress = 0;
         int64_t count = 0;
         std::unique_ptr<CCoinsViewCursor> pcursor;
+        CBlockIndex* tip;
         {
```
Comment (fqlx) [inline src/rpc/blockchain.cpp]: Can you remove unused curly braces

File: src/rpc/blockchain.cpp [hunk 3]
```diff
@@ -2293,15 +2297,20 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
             LOCK(cs_main);
             ::ChainstateActive().ForceFlushStateToDisk();
             pcursor = std::unique_ptr<CCoinsViewCursor>(pcoinsdbview->Cursor());
             assert(pcursor);
+            tip = ::ChainActive().Tip();
```
Comment (fqlx) [inline src/rpc/blockchain.cpp]: I'd just declare and initialize tip here instead of at the top and same with pcursor

**Review by fqlx — CHANGES_REQUESTED** [review]

File: src/rpc/blockchain.cpp (hunk 2 above)
Comment (promag) [reply src/rpc/blockchain.cpp]: The block is required for tcs_main scoped lock.

Comment (fqlx) [reply src/rpc/blockchain.cpp]: Can you show me this documentation? I'm not familiar with it

Comment (sipa) [reply src/rpc/blockchain.cpp]: You cannot access variables which are shared across threads without grabbing the correct lock.

**Comment by laanwj** [comment]
Concept ACK

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -2293,15 +2297,20 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
+            assert(tip);
         }
         bool res = FindScriptPubKey(g_scan_progress, g_should_abort_scan, count, pcursor.get(), needles, coins);
         result.pushKV("success", res);
-        result.pushKV("searched_items", count);
+        result.pushKV("txouts", count);
+        result.pushKV("height", tip->nHeight);
```
Comment (jonasschnelli) [inline src/rpc/blockchain.cpp]: nit: maybe `tipheight` or `bestblockheight`?
But maybe `height` is also okay.

**Review by jonasschnelli — APPROVED** [review]
utACK e36bb00f85f0f26063433c112ef675141b88ba56

File: src/rpc/blockchain.cpp [hunk 5]
```diff
@@ -2199,17 +2199,21 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
+            "    \"desc\": \"descriptor\",        (string) A specialized descriptor for the matched scriptPubKey\n"
+            "    \"amount\": x.xxx,             (numeric) The total amount in " + CURRENCY_UNIT + " of the unspent output\n"
+            "    \"height\": n,                 (numeric) Height of the unspent transaction output\n"
             "   }\n"
             "   ,...], \n"
-            " \"total_amount\" : x.xxx,          (numeric) The total amount of all found unspent outputs in " + CURRENCY_UNIT + "\n"
+            "  \"total_amount\": x.xxx,          (numeric) The total amount of all found unspent outputs in " + CURRENCY_UNIT + "\n"
             "]\n"
                 },
                 RPCExamples{""},
```
Comment (jonatack) [inline src/rpc/blockchain.cpp]: Would be good to add one or two bitcoin-cli examples here.

File: src/rpc/blockchain.cpp [hunk 6]
```diff
@@ -2199,17 +2199,21 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
-            "    \"txid\" : \"transactionid\",     (string) The transaction id\n"
-            "    \"vout\": n,                    (numeric) the vout value\n"
-            "    \"scriptPubKey\" : \"script\",    (string) the script key\n"
-            "    \"desc\" : \"descriptor\",        (string) A specialized descriptor for the matched scriptPubKey\n"
-            "    \"amount\" : x.xxx,             (numeric) The total amount in " + CURRENCY_UNIT + " of the unspent output\n"
-            "    \"height\" : n,                 (numeric) Height of the unspent transaction output\n"
+            "   {\n"
+            "    \"txid\": \"hash\",              (string) The transaction id\n"
+            "    \"vout\": n,                   (numeric) the vout value\n"
+            "    \"scriptPubKey\": \"script\",    (string) the script key\n"
```
Comment (jonatack) [inline src/rpc/blockchain.cpp]: In lines 2209-2210, s/the/The/

File: src/rpc/blockchain.cpp [hunk 7]
```diff
@@ -2199,17 +2199,21 @@ UniValue scantxoutset(const JSONRPCRequest& request)
...
+            "    \"desc\": \"descriptor\",        (string) A specialized descriptor for the matched scriptPubKey\n"
+            "    \"amount\": x.xxx,             (numeric) The total amount in " + CURRENCY_UNIT + " of the unspent output\n"
+            "    \"height\": n,                 (numeric) Height of the unspent transaction output\n"
             "   }\n"
             "   ,...], \n"
```
Comment (jonatack) [inline src/rpc/blockchain.cpp]: Trailing space can be removed here between the comma and `\n`

**Comment by jonatack** [comment]
ACK e36bb00f85f0f26063433c112ef675141b88ba56 modulo nits. Code review, compiled/ran help, inspected a few of the `scantxoutset` outputs in the functional test. Concrete CLI examples in the help would be nice to have. Could update the PR description from `result` to `success`.

**Review by jonatack — COMMENTED** [review]
ACK e36bb00 modulo nits. Code review, compiled/ran help, inspected a few of the `scantxoutset` outputs in the functional test. Concrete CLI examples in the help would be nice to have.

File: src/rpc/blockchain.cpp (hunk 5 above)
Comment (promag) [reply src/rpc/blockchain.cpp]: What do you suggest?

**Comment by laanwj** [comment]
ACK bdd6a4fd5da44c2575be9195ecb4213a13e74511

**Comment by jonatack** [comment]
> I think this is enough of an improvement in itself, examples can be added later.

Yes. I might have a look at adding examples.

**Pull request merged**